src/setfit/modeling.py

```diff
@@ -34,52 +94,281 @@ def __init__(self, model, max_seq_length: int, add_normalization_layer: bool) ->
             self.model._modules["2"] = models.Normalize()
 
 
+class SetFitHead(models.Dense):
+    """
+    A SetFit head that supports binary and multi-class classification
+    for end-to-end training.
+
+    To be compatible with Sentence Transformers, we inherit `Dense` from:
+    https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Dense.py
+
+    Args:
+        in_features (`int`, *optional*):
+            The embedding dimension from the output of the SetFit body. If `None`, defaults to `LazyLinear`.
+        out_features (`int`, defaults to `1`):
+            The number of targets.
+        temperature (`float`):
+            A logits' scaling factor when using multi-targets (i.e., number of targets more than 1).
+        bias (`bool`, *optional*, defaults to `True`):
+            Whether to add bias to the head.
+        device (`torch.device`, str, *optional*):
+            The device the model will be sent to. If `None`, will check whether GPU is available.
+    """
+
+    def __init__(
+        self,
+        in_features: Optional[int] = None,
+        out_features: int = 1,
+        temperature: float = 1.0,
+        bias: bool = True,
+        device: Optional[Union[torch.device, str]] = None,
+    ) -> None:
+        super(models.Dense, self).__init__()  # init on models.Dense's parent: nn.Module
+
+        self.linear = None
+        if in_features is not None:
+            self.linear = nn.Linear(in_features, out_features, bias=bias)
+        else:
+            self.linear = nn.LazyLinear(out_features, bias=bias)
+
+        self.in_features = in_features
+        self.out_features = out_features
+        self.temperature = temperature
+        self.bias = bias
+        self.device = device or "cuda" if torch.cuda.is_available() else "cpu"
+
+        self.to(device)
+        self.apply(self._init_weight)
+
+    def forward(
+        self, features: Union[Dict[str, torch.Tensor], torch.Tensor], temperature: Optional[float] = None
+    ) -> Union[Dict[str, torch.Tensor], torch.Tensor]:
+        """
+        SetFitHead can accept embeddings in:
+        1. Output format (`dict`) from Sentence-Transformers.
+        2. Pure `torch.Tensor`.
+
+        Args:
+            features (`Dict[str, torch.Tensor]` or `torch.Tensor):
+                The embeddings from the encoder. If using `dict` format,
+                make sure to store embeddings under the key: 'sentence_embedding'
+                and the outputs will be under the key: 'prediction'.
+            temperature (`float`, *optional*):
+                A logits' scaling factor when using multi-targets (i.e., number of targets more than 1).
+                Will override the temerature given during initialization.
+        Returns:
+        [`Dict[str, torch.Tensor]` or `torch.Tensor`]
+        """
+        is_dict = False  # whether `features` is dict or not
```

Comment:
nit: I think it helps to be explicit whenever possible 
```suggestion
        is_features_dict = False  # whether `features` is dict or not
```